transpiler/transpiler.go

```diff
@@ -0,0 +1,562 @@
+	}
+	return l + 1
+}
+
+// Errorf returns an error with the node position.
+func (o *output) Errorf(n ast.Node, format string, a ...interface{}) error {
+	l := o.findLine(int(n.Pos()))
+	return fmt.Errorf("line %d: %s\n%# v", l, fmt.Sprintf(format, a...), pretty.Formatter(n))
+}
+
+// handleDecl handles a declaration.
+//
+// It can be a function, a variable, a constant, an import, etc.
+func handleDecl(out *output, d ast.Decl) error {
+	switch decl := d.(type) {
+	case *ast.GenDecl:
+		return handleGenDecl(out, decl)
+	case *ast.FuncDecl:
+		return handleFuncDecl(out, decl)
+	default:
+		return out.Errorf(d, "unsupported decl")
+	}
+}
+
+// handleGenDecl handles a file level declaration; a constant, a variable or an
+// import statement.
+func handleGenDecl(out *output, gd *ast.GenDecl) error {
+	for _, s := range gd.Specs {
+		switch spec := s.(type) {
+		case *ast.ValueSpec:
+			//pretty.Print(spec)
+			if err := handleValueSpec(out, spec); err != nil {
+				return err
+			}
+		case *ast.ImportSpec:
+			// Ignore imports except for comments.
+			out.Writef(s, "")
+		default:
+			return out.Errorf(s, "unsupported spec")
+		}
+		// TODO(maruel): Print spacing between declarations.
+	}
+	return nil
+}
+
+func guessType(vs *ast.ValueSpec) (token.Token, string, error) {
+	if len(vs.Values) == 0 {
+		// It is an default value, e.g. "var a int". It can't be const.
+		// token.Lookup() is not very useful as it expects "STRING" instead of
+		// "string".
+		switch n := vs.Type.(*ast.Ident).Name; n {
+		case "int":
+			return token.INT, "0", nil
+		case "string":
+			return token.STRING, "\"\"", nil
+		default:
+			return token.ILLEGAL, "", fmt.Errorf("unsupported type: %s", n)
+		}
+	}
+	// Normal declaration of type "var a = 1" or "const a = 1".
+	if len(vs.Values) > 1 {
+		return token.ILLEGAL, "", fmt.Errorf("unsupported # of values: %v", vs.Names)
+	}
+	l, ok := vs.Values[0].(*ast.BasicLit)
+	if !ok {
+		return token.ILLEGAL, "", fmt.Errorf("unsupported value: %#v", vs.Values[0])
+	}
+	return l.Kind, l.Value, nil
+}
+
+func isValueConst(vs *ast.ValueSpec) bool {
+	return vs.Names[0].Obj.Kind == ast.Con
+}
+
+// handleValueSpec handles a file level a constant or variable.
+func handleValueSpec(out *output, vs *ast.ValueSpec) error {
+	if len(vs.Names) == 0 {
+		return out.Errorf(vs, "unsupported # of value names: %v", vs.Names)
+	}
+	var decl []string
+	kind, lit, err := guessType(vs)
+	if err != nil {
+		return out.Errorf(vs, "%s", err)
+	}
+	isConst := isValueConst(vs)
+	typ := tokenStr(kind, isConst)
+	if len(typ) == 0 {
+		return out.Errorf(vs, "unsupported literal kind: %s", kind)
+	}
+	// Strictly speaking the C++ version could also define all the variables on
+	// one line but the following is easier to implement.
+	for _, name := range vs.Names {
+		out.Writef(vs, "%s;\n", strings.Join(append(decl, typ, name.Name, "=", lit), " "))
+	}
+	return nil
+}
+
+func tokenStr(kind token.Token, isConst bool) string {
+	switch kind {
```

Comment:
also here, it seems we could have something common between `nodeType` `fieldToType` and `tokenStr`